### PR TITLE
log third-party reporter errors

### PR DESCRIPTION
**Pull request by adamgruber**
Fixes https://github.com/mochajs/mocha/issues/1521

### CODE REVIEW

File: lib/mocha.js [hunk 1]
```diff
@@ -141,7 +141,11 @@ Mocha.prototype.reporter = function(reporter, reporterOptions){
     reporter = reporter || 'spec';
     var _reporter;
     try { _reporter = require('./reporters/' + reporter); } catch (err) {};
-    if (!_reporter) try { _reporter = require(reporter); } catch (err) {};
+    if (!_reporter) try { _reporter = require(reporter); } catch (err) {
+      err.message.indexOf('Cannot find module') !== -1 ?
+        console.warn('"' + reporter + '" reporter not found') :
```
Comment (a8m) [inline lib/mocha.js]: IMO, it's better to stay consistent with the same code-style we have today.
e.g: [1](https://github.com/mochajs/mocha/blob/master/lib/reporters/base.js#L111-L114), [2](https://github.com/mochajs/mocha/blob/master/lib/utils.js#L441-L442)

Comment (adamgruber) [inline lib/mocha.js]: That's fair. I updated the code.

**Comment by dasilvacontin** [comment]
Thanks for the PR. Please squash commits.

Nice catch @a8m!

**Comment by adamgruber** [comment]
@boneskull What about this needs work?

**Comment by dasilvacontin** [comment]
Maybe it's because you didn't say you were done squashing, or what is it, @boneskull?

**Comment by danielstjules** [comment]
Looks like this may have been forgotten after the PR was cleaned up. I'm gonna go ahead and rebase/merge this in since it's a small fix, and really helpful for anyone developing third party reporters :)

The style of the change is consistent with the surrounding code, but could probably be revisited later if/when JSHint gets integrated in the project & contribution guidelines. In particular, I'm not sure that having a single line if statement lead to a multi-line catch block is a great idea, but that's a nitpick.

**Comment by danielstjules** [comment]
Before PR
``` javascript
$ cat node_modules/example_reporter/index.js
{{{ohno! invalid js

$ ./bin/mocha --reporter example_reporter

/Users/danielstjules/git/mocha/node_modules/example_reporter/index.js:1
function (exports, require, module, __filename, __dirname) { {{{ohno! invalid
                                                                    ^
Error: invalid reporter "example_reporter"
    at Mocha.reporter (/Users/danielstjules/git/mocha/lib/mocha.js:152:27)
    at Object.<anonymous> (/Users/danielstjules/git/mocha/bin/_mocha:197:7)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:901:3

$ ./bin/mocha --reporter test

/Users/danielstjules/git/mocha/lib/mocha.js:152
    if (!_reporter) throw new Error('invalid reporter "' + reporter + '"');
                          ^
Error: invalid reporter "test"
    at Mocha.reporter (/Users/danielstjules/git/mocha/lib/mocha.js:152:27)
    at Object.<anonymous> (/Users/danielstjules/git/mocha/bin/_mocha:197:7)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:901:3
```

After PR
``` javascript
$ ./bin/mocha --reporter example_reporter

/Users/danielstjules/git/mocha/node_modules/example_reporter/index.js:1
function (exports, require, module, __filename, __dirname) { {{{ohno! invalid
                                                                    ^
"example_reporter" reporter blew up with error:
SyntaxError: Unexpected token !
    at Module._compile (module.js:439:25)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Mocha.reporter (/Users/danielstjules/git/mocha/lib/mocha.js:147:39)
    at Object.<anonymous> (/Users/danielstjules/git/mocha/bin/_mocha:197:7)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:901:3
Error: invalid reporter "example_reporter"
    at Mocha.reporter (/Users/danielstjules/git/mocha/lib/mocha.js:156:27)
    at Object.<anonymous> (/Users/danielstjules/git/mocha/bin/_mocha:197:7)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:901:3

$ ./bin/mocha --reporter test
"test" reporter not found

/Users/danielstjules/git/mocha/lib/mocha.js:156
    if (!_reporter) throw new Error('invalid reporter "' + reporter + '"');
                          ^
Error: invalid reporter "test"
    at Mocha.reporter (/Users/danielstjules/git/mocha/lib/mocha.js:156:27)
    at Object.<anonymous> (/Users/danielstjules/git/mocha/bin/_mocha:197:7)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:901:3
```

Note that in the second case, it mentions that the reporter couldn't be found twice, which is redundant. I'm gonna submit a new PR that avoids this:

```
$ ./bin/mocha --reporter test
"test" reporter not found

/Users/danielstjules/git/mocha/lib/mocha.js:156
    if (!_reporter) throw new Error('invalid reporter "' + reporter + '"');
                          ^
Error: invalid reporter "test"
...
```

**Comment by danielstjules** [comment]
Merged in https://github.com/mochajs/mocha/commit/ca84810498fe1b32d2ab237b1f0b98db4a9f9da5 :) Ignore the comment I deleted - realized that having both messages makes most sense.

**Pull request closed without merging**